### PR TITLE
Track Unseen Files [#PLAT-639]

### PR DESCRIPTION
**Pull request by erinspace**
## Purpose

If a user has previously seen a file, and another user uploads a new version of that file, we'd like some indication that the new version is new! So, make the filename bold for new versions of previously viewed files.

screen shot 2018-04-17 at 10 47 56 am

## Changes
- Migrations and models for Version - User relationships
- When `create_waterbutler_log` is called and the action is in `DOWNLOAD_ACTIONS`, create a new relationship between the version accessed and the user, store it in `seen_by` on the osfstorage file version
- Return `latestVersionSeen` in `osfstorage_get_children` so that waterbutler can then also pass along this information when rendering the file trees
- Update existing tests for wb callbacks 
- Add tests to make sure relationship is created when file is viewed

## QA Notes
The QA on this one will be interesting! No API changes tho.

It'll work best with 2 users, I'll call them Gertrude and Sally.

1. Sally and Gertrude can both see a project -- they can be collaborators or not
2. Sally uploads a file to the project.
3.  Gertrude goes to the project, sees nothing different about anything
4. Gertrude clicks on the file to go to the file detail page -- she is now aware of the file's existence
5. Sally uploads a new version of the file
6. Gertrude goes back to either the project page or the main file overview page
7. Gertrude will notice that the file she had seen before now has its **filename in bold**
8. Gertrude views the file detail page for that new file version
9. Gertrude goes back to the project overview/file overview page, filename is no longer bold

Notes:
- this also works for non-contributors on projects
- the bold filename should show up on project page, file detail page, and file overview page

## Documentation

- No changes to the API,  internally tried to update new/existing methods with docstrings

## Side Effects
Now sending and receiving new metadata to WB! Shouldn't really affect anything tho.

Here's the blocking WB PR:
https://github.com/CenterForOpenScience/waterbutler/pull/337

## Ticket
Platform ticket:
https://openscience.atlassian.net/browse/PLAT-639
blocking WB ticket:
https://openscience.atlassian.net/browse/SVCS-689

### CODE REVIEW

File: addons/base/views.py [hunk 1]
```diff
@@ -309,24 +309,33 @@ def get_auth(auth, **kwargs):
 ])
 
 
+def mark_file_version_as_seen(user, path):
+    file_to_update = OsfStorageFile.objects.get(_id=path)
+    # TODO - once the version is available in the waterbutler callback, update a specific version
```
Comment (Johnetordoff) [inline addons/base/views.py]: I think with [this commit](https://github.com/CenterForOpenScience/waterbutler/commit/81bec1ad3602177bca177a1ca0e8c93adc65b5e4) we can retrieve version info via query params used the request url. But It's a judgement call as to whether this is really in scope or not because a don't think anything about viewing intermediate versions is actually part of the ticket.

In either case, I have to grab the version this way as part of the upcoming analytics so one of is going to be piggybacking off the other.

**Comment by Johnetordoff** [comment]
✅ LGTM!

File: osf/models/files.py [hunk 2]
```diff
@@ -672,6 +672,14 @@ def restore(self, recursive=True, parent=None, save=True, deleted_on=None):
         return tf
 
 
+class FileVersionUserMetadata(BaseModel):
+    user = models.ForeignKey('OSFUser')
```
Comment (caseyrollins) [inline osf/models/files.py]: Drive-by: need to specify `on_delete` for `ForeignKey` so we don't get [deprecation warnings]( https://docs.djangoproject.com/en/2.0/internals/deprecation/#deprecation-removed-in-2-0).

Comment (caseyrollins) [reply osf/models/files.py]: @erinspace, on second thought, I don't think this model is necessary. Django [creates a through table for you](https://docs.djangoproject.com/en/1.11/ref/models/fields/#django.db.models.ManyToManyField.through) when you define a M2M (`seen_by` on `FileVersion`). You would only define your own intermediary table when there's additional data to store.

Comment (sloria) [reply osf/models/files.py]: It's not strictly necessary to have the through table, but it's more explicit to have it and also `created` and `modified` fields, which can be useful metadata for debugging.

Furthermore, explicitly defining the through table is more future-proof--it's easier to add it now than it would be in the future (after rows have been added to the implicit table).

File: addons/osfstorage/views.py [hunk 3]
```diff
@@ -220,9 +224,42 @@ def osfstorage_get_children(file_node, **kwargs):
...
+                LIMIT 1
+              )
+            ) SEEN_FILE ON TRUE
+            LEFT JOIN LATERAL (
+                SELECT CASE WHEN SEEN_FILE.exists
+                THEN
+                    CASE WHEN EXISTS(
+                        SELECT (1) FROM osf_osfuser
+                        INNER JOIN osf_fileversionusermetadata ON (osf_osfuser.id = osf_fileversionusermetadata.user_id)
+                        INNER JOIN osf_guid ON (
```
Comment (sloria) [inline addons/osfstorage/views.py]: This `JOIN` duplicates the one in the query above. It might be more efficient to get the user's PK once at the top of the function. That should also simplify the queries here.

```python
user_pk = OSFUser.objects.filter(guids___id=user_id).values_list('pk', flat=True).get()
with connection.cursor() ....:
```

File: addons/osfstorage/views.py [hunk 4]
```diff
@@ -58,6 +58,7 @@ def osfstorage_update_metadata(node_addon, payload, **kwargs):
                 "version": "2",
                 "downloads": "1",
                 "checkout": "...",
+                "latestVersionSeen": {"userId": "abc12", "seen": true},
```
Comment (sloria) [inline addons/osfstorage/views.py]: I'm not sure this belongs. I think this doc is referring to the payload from WB.

Comment (erinspace) [reply addons/osfstorage/views.py]: Yep, the waterbutler payload [changed with the related PR](https://github.com/CenterForOpenScience/waterbutler/blob/develop/waterbutler/providers/osfstorage/metadata.py#L105) to include `latestVersionSeen`

Comment (sloria) [reply addons/osfstorage/views.py]: Ah ok. Then this is correct. Thanks.
